src/common/lib/transport/connectionmanager.ts

```diff
@@ -1508,129 +1261,160 @@ class ConnectionManager extends EventEmitter {
   }
 
   /**
-   * There are three stages in connecting:
-   * - preference: if there is a cached transport preference, we try to connect
-   *   on that. If that fails or times out we abort the attempt, remove the
-   *   preference and fall back to base. If it succeeds, we try upgrading it if
-   *   needed (will only be in the case where the preference is xhrs and the
-   *   browser supports ws).
-   * - base: we try to connect with the best transport that we think will
-   *   never fail for this platform. If it doesn't work, we try fallback hosts.
-   * - upgrade: given a connected transport, we see if there are any better
-   *   ones, and if so, try to upgrade to them.
-   *
-   * connectImpl works out what stage you're at (which is purely a function of
-   * the current connection state and whether there are any stored preferences),
-   * and dispatches accordingly. After a transport has been set pending,
-   * tryATransport calls connectImpl to see if there's another stage to be done.
+   * connectImpl works out which transport to initially attempt a connection with
    * */
-  connectImpl(transportParams: TransportParams, connectCount?: number): void {
+  connectImpl(transportParams: TransportParams): void {
     const state = this.state.state;
-
-    if (state !== this.states.connecting.state && state !== this.states.connected.state) {
+    if (state !== this.states.connecting.state) {
       /* Only keep trying as long as in the 'connecting' state (or 'connected'
        * for upgrading). Any operation can put us into 'disconnected' to cancel
        * connection attempts and wait before retrying, or 'failed' to fail. */
       Logger.logAction(
         Logger.LOG_MINOR,
         'ConnectionManager.connectImpl()',
-        'Must be in connecting state to connect (or connected to upgrade), but was ' + state
-      );
-    } else if (this.pendingTransports.length) {
-      Logger.logAction(
-        Logger.LOG_MINOR,
-        'ConnectionManager.connectImpl()',
-        'Transports ' + this.pendingTransports[0].toString() + ' currently pending; taking no action'
+        'Must be in connecting state to connect, but was ' + state
       );
-    } else if (state == this.states.connected.state) {
-      this.upgradeIfNeeded(transportParams);
-    } else if (this.transports.length > 1 && this.getTransportPreference()) {
-      this.connectPreference(transportParams, connectCount);
+      return;
+    }
+
+    const transportPreference = this.getTransportPreference();
+
+    if (
+      (transportPreference && transportPreference === this.baseTransport) ||
+      (this.baseTransport && !this.webSocketTransport)
```

Comment:
I don't think it's right that we're now just never unpersisting the transport preference. If you ever go to a corporate network that breaks websockets, and you end up with an xhr_polling persisted preference, that's it's no more websockets for you ever, in this current instance of the lib or any future ones that share its localstorage.

Not sure what the best solution is. maybe something like, in `connectImpl()`, if the transport preference is the base transport but the ws transport is theoretically available, kick off a immediate ws connectivity check in parallel with the base connection sequence, that removes the persisted transport preference (and unsets `wsCheckResult`) if it succeeds?

Reply:
Was thinking about this, if we perform a ws connectivity check in `connectImpl` whenever there's a base transport preference and the connectivity check succeeds before the base transport is connected, should we just abandon the base transport attempt and try websocket immediately, or just unpersist the transport preference for next time? I'm leaning towards the former